crates/subspace-networking/src/node_runner.rs

```diff
@@ -411,6 +422,23 @@ where
                 };
                 debug!("Connection closed with peer {peer_id} [{num_established} from peer]");
 
+                // TODO: Workaround for https://github.com/libp2p/rust-libp2p/discussions/3418
+                {
+                    match self.established_connections.entry((peer_id, endpoint)) {
+                        Entry::Vacant(_) => {
+                            // Nothing to do here, we are not aware of the connection being closed
+                            return;
+                        }
+                        Entry::Occupied(mut entry) => {
+                            let value = entry.get_mut();
```

Comment:
Your other change handles it in a way, but can the sem shrink()  be called only if we have an entry here?

Reply:
We only expand when new connection is established, which in turn means we should only shrink if the same connection is closed. If we didn't see closed connection ever opening, we don't shrink.

Reply:
sorry missed the return. For my understanding, when will we have count > 1 for the same `(PeerId, ConnectedPoint)`. Is this some race in the up/down event sequence?

Reply:
No, this is a single-threaded sort of even loop, it can't fire disconnect before connection (at least I'm surely hope so, it would be extremely surprising bug if it did).